### PR TITLE
PLANET-7214 Build a Posts List block using Query Loop

### PR DESCRIPTION
**Pull request by Osong-Michael**
**Ref: [PLANET-7214](https://jira.greenpeace.org/browse/PLANET-7214)**

**Testing:**

### CODE REVIEW

File: src/Loader.php [hunk 1]
```diff
@@ -157,6 +160,32 @@ public static function add_blocks(): void
         new Blocks\GuestBook();//NOSONAR
     }
 
+      /**
+     * Load editorIndex file for beta blocks.
+     */
+    public static function add_beta_blocks(): void
+    {
+        if (!BetaBlocks::is_active() && !AllowAllBlocks::is_active()) {
```
Comment (comzeradd) [inline src/Loader.php:168]: I think it makes more sense to check for Planet4Blocks flag instead of AllBlocks (in addition to Beta)

Comment (Osong-Michael) [reply src/Loader.php:168]: We need the `All Blocks` Feature enabled so we have access to the `Query Loop Block` which the `Posts List block` extends, if not, it does not show up in the editor.

Now that I think about it, is it redundant to have the check for `AllBlocks` here, because it still won't show up if that is not enabled but `Beta Blocks` is 🤔 ?

File: assets/src/scss/blocks/PostsListBlock.scss [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  .wp-block-group.is-layout-flow {
+    flex-basis: 70%;
+  }
+
+  .wp-block-post-title {
+    font-weight: 700;
+    color: var(--grey-900);
+    font-family: var(--font-family-tertiary);
+    font-size: 1.375rem;
+    margin-top: 5px;
```
Comment (GP-Dan-Tovbein) [inline assets/src/scss/blocks/PostsListBlock.scss:35]: Can we use spacing variables? i.e. `$sp-1x` which is equal to `6px` instead of `5px`. The we have all the spacing values already standardized.

File: assets/src/scss/blocks/PostsListBlock.scss [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  .wp-block-post-excerpt {
+    margin-top: -15px;
```
Comment (GP-Dan-Tovbein) [inline assets/src/scss/blocks/PostsListBlock.scss:39]: Same here!

File: assets/src/scss/blocks/PostsListBlock.scss [hunk 4]
```diff
@@ -0,0 +1,143 @@
+.posts-list-query {
+  .wp-block-post-template.is-layout-flow li {
+    width: 100% !important;
+  }
+
+  .wp-block-columns.is-layout-flex {
+    gap: 20px;
```
Comment (GP-Dan-Tovbein) [inline assets/src/scss/blocks/PostsListBlock.scss:7]: Same here!

**Review by GP-Dan-Tovbein — COMMENTED** [review]
General commens! 
1. Can we use spacing variables?
2. Can we check the places where we use `font-size|weight` or `line-height` could be replaced by tokens?

File: assets/src/blocks/PostsList/PostsListBlock.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    },
+    scope: ['inserter'],
+    innerBlocks: [
+      ['core/heading', {content: 'Related Posts'}],
+      ['core/paragraph', {
+        placeholder: __('Enter description', 'planet4-blocks-backend'),
+        style: {
+          spacing: {
+            margin: {
+              top: '25px',
```
Comment (GP-Dan-Tovbein) [inline assets/src/blocks/PostsList/PostsListBlock.js:42]: Why don't we use spacing values?

File: assets/src/scss/blocks/PostsListBlock.scss [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+    img {
+      @include medium-and-up {
+        height: 180px;
+      }
+
+      @include large-and-up {
+        height: 210px;
+      }
+      width: 100% !important;
```
Comment (GP-Dan-Tovbein) [inline assets/src/scss/blocks/PostsListBlock.scss:21]: Could we move the next properties to line 14?

File: assets/src/scss/blocks/PostsListBlock.scss [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+      object-fit: cover;
+    }
+  }
+
+  .wp-block-group.is-layout-flow {
+    flex-basis: 70%;
+  }
+
+  .wp-block-post-title {
+    font-weight: $font-weight-bold;
```
Comment (GP-Dan-Tovbein) [inline assets/src/scss/blocks/PostsListBlock.scss:31]: You can use tokens instead!

File: assets/src/scss/blocks/PostsListBlock.scss [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    color: var(--grey-900);
+    font-family: var(--font-family-tertiary);
+    font-size: $font-size-l--font-family-primary;
```
Comment (GP-Dan-Tovbein) [inline assets/src/scss/blocks/PostsListBlock.scss:34]: I think it's better to use `CSS variables` instead!

File: assets/src/scss/blocks/PostsListBlock.scss [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+  .is-layout-flow > .is-layout-flow {
+    display: flex;
+    gap: $sp-x;
+    font-size: calc($font-size-xxs--font-family-primary - 2px);
+    color: var(--grey-600);
+    font-family: var(--font-family-paragraph-secondary);
+  }
+
+  .wp-block-post-author-name {
+    font-size: calc($font-size-xxs--font-family-primary - 2px);
```
Comment (GP-Dan-Tovbein) [inline assets/src/scss/blocks/PostsListBlock.scss:75]: This looks hard to follow up!

File: assets/src/scss/blocks/PostsListBlock.scss [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    &::after {
+      content: ".";
+      font-size: calc($font-size-m--font-family-primary + 5px);
+      margin-top: calc(2px - $sp-1);
+      margin-left: $sp-x;
+    }
+  }
+
+  .wp-block-post-terms {
+    margin-top: calc(3px - $sp-1);
```
Comment (GP-Dan-Tovbein) [inline assets/src/scss/blocks/PostsListBlock.scss:94]: Same as [here](https://github.com/greenpeace/planet4-master-theme/pull/2137/files#r1379129875)

File: assets/src/scss/editorOverrides.scss [hunk 11]
```diff
@@ -415,3 +418,48 @@ input.describe[type=text][data-setting=caption] {
     }
   }
 }
+
+// Posts List Block Override
+.posts-list-query {
+  .wp-block-post-template.is-layout-flow li {
+    height: 240px;
+    margin-bottom: calc($sp-1 + 2px);
```
Comment (GP-Dan-Tovbein) [inline assets/src/scss/editorOverrides.scss:426]: Same as [here](https://github.com/greenpeace/planet4-master-theme/pull/2137/files#r1379129875)

File: src/Loader.php [hunk 12]
```diff
@@ -157,6 +159,42 @@ public static function add_blocks(): void
         new Blocks\GuestBook();//NOSONAR
     }
 
+    /**
+     * Load editorIndex file and add core Query Loop Block for Beta Blocks.
+     */
+    public static function add_beta_blocks(): void
+    {
+        if (!BetaBlocks::is_active()) {
```
Comment (GP-Dan-Tovbein) [inline src/Loader.php:167]: I think we can move this statement to [here](https://github.com/greenpeace/planet4-master-theme/pull/2137/files#diff-1314d7908e6167b9fbe3d52c6f0c3cadf76637c75eca9c9c36d66106c579edd3R62). And we'll avoid to invoke unnecessary this function.

File: assets/src/blocks/PostsList/PostsListBlock.js [hunk 13]
```diff
@@ -0,0 +1,68 @@
+export const POSTS_LIST_BLOCK = 'planet4-blocks/posts-list-block';
+
+export const registerPostsListBlock = () => {
+  const {registerBlockVariation} = wp.blocks;
+  const {__} = wp.i18n;
+
+  return registerBlockVariation('core/query', {
+    name: POSTS_LIST_BLOCK,
+    title: 'P4 Posts List Block',
```
Comment (GP-Dan-Tovbein) [inline assets/src/blocks/PostsList/PostsListBlock.js:9]: Don't we need to translate this text?

File: assets/src/scss/blocks/PostsListBlock.scss (hunk 7 above)
Comment (Osong-Michael) [reply assets/src/scss/blocks/PostsListBlock.scss:31]: That is a token from [here](https://github.com/greenpeace/planet4-master-theme/blob/main/assets/src/scss/base/_tokens.scss#L114), I can use the css variable instead if that is what you mean.

File: assets/src/scss/blocks/PostsListBlock.scss (hunk 9 above)
Comment (Osong-Michael) [reply assets/src/scss/blocks/PostsListBlock.scss:75]: Based on the tokens we have, we don't have a token for `14px`, so I used the `calc()` to subtract `2px` from the token we have for `16px` (based of your recommendation to use tokens). I could add the values in comments to make it clearer.

File: assets/src/scss/blocks/PostsListBlock.scss (hunk 10 above)
Comment (Osong-Michael) [reply assets/src/scss/blocks/PostsListBlock.scss:94]: I am using [spacing values](https://github.com/greenpeace/planet4-master-theme/blob/main/assets/src/scss/base/_variables.scss#L14) here, just that when I need specific units I need to calculate them, like above I need a negative value to move the element up.

File: assets/src/blocks/PostsList/PostsListBlock.js (hunk 13 above)
Comment (comzeradd) [reply assets/src/blocks/PostsList/PostsListBlock.js:9]: I think we don't offer translation for our block names, just to make things less confusing on Handbook documentation.

Comment (comzeradd) [reply assets/src/blocks/PostsList/PostsListBlock.js:9]: Can't we just use "Posts List" as a name?

Comment (Osong-Michael) [reply assets/src/blocks/PostsList/PostsListBlock.js:9]: I changed the name because WP offers a variation of the Query Loop which is also called Posts List, so we may need to change ours or add `Block` at the end to differentiate the two.

Screenshot 2023-11-02 at 09 42 04

File: assets/src/blocks/PostsList/PostsListBlock.js (hunk 5 above)
Comment (GP-Dan-Tovbein) [reply assets/src/blocks/PostsList/PostsListBlock.js:42]: For instance `24px`

File: assets/src/blocks/PostsList/PostsListBlock.js [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+      ['core/paragraph', {
+        placeholder: __('Enter description', 'planet4-blocks-backend'),
+        style: {
+          spacing: {
+            margin: {
+              top: '25px',
+              bottom: '35px',
+            },
+          },
+        },
```
Comment (GP-Dan-Tovbein) [inline assets/src/blocks/PostsList/PostsListBlock.js:46]: And here `36px`

Comment (GP-Dan-Tovbein) [reply assets/src/blocks/PostsList/PostsListBlock.js:46]: But i'd suggest to validate with Houssam first!

**Review by GP-Dan-Tovbein — COMMENTED** [review]
Code looks good to me!

General comment! From my point of view, I disagree to use `calc()` with `subract` or `add` to `margin` or `padding` since it's really hard to understand the assigned value.
In that cases I'd prefer to follow up with the design team first and then define static values, such as `32px`. It should be always multiple of `4`.

File: assets/src/scss/style.scss [hunk 15]
```diff
@@ -79,6 +79,9 @@ Text Domain: planet4-master-theme
 // CSS variables
 @import "base/css-variables";
 
+// Posts List Block
+@import "blocks/PostsListBlock";
```
Comment (mleray) [inline assets/src/scss/style.scss:83]: Why do we not add this to the [blocks style file](https://github.com/greenpeace/planet4-master-theme/blob/main/assets/src/blocks/styles/blocks.scss) instead?

Comment (Osong-Michael) [reply assets/src/scss/style.scss:83]: Oh my bad, I had started working on this before Sagar merged the code to move blocks to the master theme, so let me change that.

**Review by mleray — COMMENTED** [review]
We can change this description via the `description` attribute, maybe we should? It's a bit vague right now since it's the default

Screenshot 2023-11-13 at 10 28 05

File: assets/src/scss/style.scss (hunk 15 above)
Comment (mleray) [reply assets/src/scss/style.scss:83]: Maybe it actually makes more sense to do it this way 🤔 I'm not sure we need a separate style file for blocks, also it seems strange to have some styles outside of the `scss` folder

**Comment by Osong-Michael** [comment]
> PLANET-7214


> We can change this description via the `description` attribute, maybe we should? It's a bit vague right now since it's the default
> 
> Screenshot 2023-11-13 at 10 28 05

Tagging @comzeradd to see if he has any suggestions.

**Comment by comzeradd** [comment]
> We can change this description via the `description` attribute, maybe we should?

We can adapt the one suggested by the design team for the Articles block:

Posts List is the place in Planet 4 that the latest articles, press releases and publications can be found.

**Review by mleray — COMMENTED** [review]
I think that we should remove most of the sidebar options, that might be confusing. According to the ticket requirements we only need to be able to filter by taxonomies, so adding `allowedControls: ['taxQuery']` should do the trick 🙂 See [docs](https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/extending-the-query-loop-block/#disabling-irrelevant-or-unsupported-query-controls) for more details

**Review by mleray — COMMENTED** [review]
I see in the ticket requirements that it should be possible to change the amount of posts, I've added something similar in my [Actions List PR](https://github.com/greenpeace/planet4-master-theme/pull/2149/files#diff-aabb623e32776e19ec89d1c093007def03e230b12831fa51b20ebcfa735a5513) if you want inspiration (although that PR is still a draft so if you have better ideas of how to implement this, please let me know!)

File: src/Loader.php (hunk 12 above)
Comment (mleray) [reply src/Loader.php:167]: I think that here we should also check that the blocks plugin is active, by adding `!Planet4Blocks::is_active()`

**Comment by Osong-Michael** [comment]
> I see in the ticket requirements that it should be possible to change the amount of posts, I've added something similar in my [Actions List PR](https://github.com/greenpeace/planet4-master-theme/pull/2149/files#diff-aabb623e32776e19ec89d1c093007def03e230b12831fa51b20ebcfa735a5513) if you want inspiration (although that PR is still a draft so if you have better ideas of how to implement this, please let me know!)

Yes, I discussed that with Nikos while you were away and we decided to leave that option in the toolbar above the block since we may soon allow Editors to use the _**Query Loop Block**_ and thats's where the option to change the number of posts is.

File: src/Loader.php (hunk 12 above)
Comment (Osong-Michael) [reply src/Loader.php:167]: > I think that here we should also check that the blocks plugin is active, by adding `!Planet4Blocks::is_active()` (like we do in the `add_blocks` function)

I am not sure, maybe @comzeradd can shed more light on this, but my understanding was this implementation was to exist despite us having to move the _Blocks_ to the theme folder, so the _Beta Blocks_ should be implemented even without the Planet4Blocks feature

Comment (comzeradd) [reply src/Loader.php:167]: For now I think it's fine to just check for Beta blocks. Since this feature flag only exists on dev sites there is no danger for someone accidentally enabling this block in production.

Once we are ready to actual get this block live then we can substitute the Beta check with the `Planet4Blocks` feature flag.

**Comment by comzeradd** [comment]
> Yes, I discussed that with Nikos while you were away and we decided to leave that option in the toolbar above the block

Yes, we still need to [limit to 3](https://github.com/greenpeace/planet4-master-theme/pull/2137/files#diff-ce4b283c7f18ddd76b48c753fd7426e195c1eae9a2305cf9fdeb56bdbf30782aR22) by default and allow editors to change that but since QLB offers that functionality out of the box in its toolbar we can use just that instead of replicating it in the sidebar.

File: src/Loader.php (hunk 12 above)
Comment (mleray) [reply src/Loader.php:167]: Maybe we can directly add this code to the `add_blocks` function? Instead of creating a new `add_beta_blocks`? 🤔

Comment (mleray) [reply src/Loader.php:167]: Also, we already enqueue this script [in the MasterBlocks file](https://github.com/greenpeace/planet4-master-theme/blob/1200cf93530ee9a9e0a85569644ac74617971d2f/src/MasterBlocks.php#L51), should we remove that then if we add it here?

File: assets/src/blocks/PostsList/PostsListBlock.js [hunk 16]
```diff
@@ -0,0 +1,70 @@
+export const POSTS_LIST_BLOCK = 'planet4-blocks/posts-list-block';
+
+export const registerPostsListBlock = () => {
+  const {registerBlockVariation} = wp.blocks;
+  const {__} = wp.i18n;
+
+  return registerBlockVariation('core/query', {
```
Comment (mleray) [inline assets/src/blocks/PostsList/PostsListBlock.js:7]: If you add `category: 'planet4-blocks-beta'` then it will show in the editor in the blocks list along with our other beta blocks, which I believe is what we want?

Screenshot 2023-11-16 at 11 00 26

Comment (mleray) [reply assets/src/blocks/PostsList/PostsListBlock.js:7]: (not sure if we should add `(beta`) to the name 🤔)

File: src/Loader.php (hunk 12 above)
Comment (Osong-Michael) [reply src/Loader.php:167]: Hey @mleray, thanks for the feedback.
So just a few points to see what you think. The reason for creating the `add_beta_blocks` function is so that in the future if we have to change things we can just isolate that particular code and integrate it into the `add_blocks`, depending on the route we choose. Also, it is easier to manage the script for the Beta blocks since we are enqueueing just that script while the `MasterBlock` file does a few other things.

Comment (mleray) [reply src/Loader.php:167]: But won't we be loading the file twice then? 🤔

Comment (Osong-Michael) [reply src/Loader.php:167]: I think the file is only loaded when the `Planet4Blocks` feature is enabled, which by default is not since that will throw an error because of the GuestBook block

Comment (mleray) [reply src/Loader.php:167]: Actually I think most of our sites have it enabled, no? Doesn't it mean that our blocks plugin is enabled?

Comment (Osong-Michael) [reply src/Loader.php:167]: Let me double check again.

Comment (Osong-Michael) [reply src/Loader.php:167]: My bad, just realised that. I will have the code updated asap

File: assets/src/blocks/PostsList/PostsListBlock.js [hunk 17]
```diff
@@ -0,0 +1,71 @@
+const POSTS_LIST_BLOCK = 'planet4-blocks/posts-list-block';
+
+export const registerPostsListBlock = () => {
+  const {registerBlockVariation} = wp.blocks;
+  const {__} = wp.i18n;
+
+  return registerBlockVariation('core/query', {
+    name: POSTS_LIST_BLOCK,
+    title: 'Posts List',
```
[the remark this answers is not in the document]
Comment (comzeradd) [reply assets/src/blocks/PostsList/PostsListBlock.js:9]: I think it’s ok for now since it’s Beta. I’ll open a follow up ticket to investigate hiding the native QL block variations

**Review by mleray — COMMENTED** [review]
I think we need to add a limit to the amount of lines for the headline, I believe it should be 2?

Screenshot 2023-11-22 at 13 48 58

File: assets/src/blocks/PostsList/PostsListBlock.js [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+          ['core/group', {}, [
+            ['core/post-terms', {
+              term: 'category',
+              separator: ' ',
+            }],
+            ['core/post-title', {isLink: true, linkTarget: '_blank'}],
+            ['core/post-excerpt'],
+            ['core/group', {className: 'posts-list-meta'}, [
+              ['core/post-author-name', {isLink: true, linkTarget: '_blank'}],
+              ['core/post-date'],
```
Comment (mleray) [inline assets/src/blocks/PostsList/PostsListBlock.js:63]: Do we need to add the reading time after this, similar to what we do for [listing pages](https://github.com/greenpeace/planet4-master-theme/blob/05346596a9e9255c5df800429c559550b68d6800/parts/query-grid.html#L29)?

Comment (Osong-Michael) [reply assets/src/blocks/PostsList/PostsListBlock.js]: Hey @mleray and @comzeradd, so I was looking into this and after a conversation with @lithrel, this [block](https://github.com/greenpeace/planet4-master-theme/blob/main/functions.php#L237) is not registered on the frontend so it throws an error and crashed the editor. We could create another ticket to register the block to the frontend and add to the Posts List block(so as to maintain this ticket's scope) or do we do it in this ticket?

Comment (comzeradd) [reply assets/src/blocks/PostsList/PostsListBlock.js]: I can create a follow up ticket for the "reading time" block to be added in the Posts List

File: assets/src/blocks/PostsList/index.js [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+      }],
+      ['core/post-template', {}, [
+        ['core/columns', {}, [
+          ['core/post-featured-image', {isLink: true, linkTarget: '_blank'}],
+          ['core/group', {}, [
+            ['core/post-terms', {
+              term: 'category',
+              separator: ' ',
+            }],
+            ['core/post-title', {isLink: true, linkTarget: '_blank'}],
```
Comment (mleray) [inline assets/src/blocks/PostsList/index.js:59]: I'm wondering if these links should actually have `_blank` target? The Articles block doesn't do that 🤔 I did the same thing for the Actions List, and that might have been a mistake since the Covers block also doesn't do that 😬

Comment (comzeradd) [reply assets/src/blocks/PostsList/index.js:59]: Ah yes, I don't think these links are supposed to open in a new tab in either of these blocks :)

Comment (Osong-Michael) [reply assets/src/blocks/PostsList/index.js:59]: I will update the Actions List block too

File: assets/src/scss/editorOverrides.scss [hunk 20]
```diff
@@ -415,3 +417,48 @@ input.describe[type=text][data-setting=caption] {
     }
   }
 }
+
+// Posts List Block Override
+.posts-list-query {
```
[the remark this answers is not in the document]
Comment (Osong-Michael) [reply assets/src/scss/editorOverrides.scss:422]: @mleray, yes we do, some elements appear in the editor that we do not want to show to editors, so I hide them using css, also the styling is a bit different in the editor than on the page, so need this to make it similar.

**Review by mleray — APPROVED** [review]
Looks good to me 👍 Thanks for your patience Osong! 🙏
